Refactor UA Overview to support step-completion

cjcenizal opened the pull request:
This enables future work to address https://github.com/elastic/kibana/issues/110737 by storing step-completion state at the root of the Overview page, and delegating step-completion logic to each step component.

When each step component detects that it's completed or not completed, it calls a callback function that's injected as a prop. This callback function sets the step completion state in the Overview component, which is then passed back down to the step "provider" function, so that it can set a `complete` or `incomplete` state.

I think this indirection is necessary, due to the way `EuiSteps` is designed -- it's configured with a config object instead of React components, so we need to store this state at the root instead of inside each step "provider", e.g. `backup_step.tsx`. But maybe I'm missing something!

@alisonelizabeth @sabarasaba Do you think this pattern is the right approach?

Partially addresses https://github.com/elastic/kibana/issues/110737 (doesn't implement new success state design for the issues)
Fixes https://github.com/elastic/kibana/issues/111255

elasticmachine commented:
Pinging @elastic/kibana-stack-management (Team:Stack Management)

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151151/)
* Commit: 48720733eb7d87ae23ccba82c2dd0ee161a87a59
* Storybooks not built yet
* This comment will update when the build is complete

### Failed CI Steps
* [Check Types and Public API Docs](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151151/execution/node/428/log)

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151151/","number":"151151","commit":"48720733eb7d87ae23ccba82c2dd0ee161a87a59"}]}
PIPELINE-->

sabarasaba commented:
I think should work just fine, I had very similar approach in mind for tackling this particular problem 👍🏼

sabarasaba commented on x-pack/plugins/upgrade_assistant/public/application/components/overview/overview.tsx:
```diff
@@ -97,7 +107,11 @@ export const Overview: FunctionComponent = () => {
+            getBackupStep({
+              cloud,
+              isComplete: isStepComplete('backup'),
+              setIsComplete: setCompletedStep.bind(null, 'backup'),
```
we probably need to memoize this, otherwise if this is added into a dependency from a useEffect it will be called ∞ 🤔

cjcenizal replied:
Seems fine now that it's not used as a dependency here: https://github.com/elastic/kibana/pull/111243/files#diff-c79c9d11f20a53d4cae48f26e280ee0ae4c7295aae8a0aebfa189df120f6090cR55

alisonelizabeth replied:
If it's memoized, will we we not need to add `// eslint-disable-next-line react-hooks/exhaustive-deps`?

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151310/)
* Commit: 7f9becd3b4bddb7ab79f388a15009a24e63ebd50
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-111243/7f9becd3b4bddb7ab79f388a15009a24e63ebd50)
* This comment will update when the build is complete

### Failed CI Steps
* [Execute xpack-kibana-ciGroup6](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151310/execution/node/718/log)
* [Execute xpack-kibana-ciGroup6](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151310/execution/node/853/log)

### History
* :broken_heart: [Build #151151](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151151/) failed 48720733eb7d87ae23ccba82c2dd0ee161a87a59

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151151/","number":"151151","commit":"48720733eb7d87ae23ccba82c2dd0ee161a87a59"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151310/","number":"151310","commit":"7f9becd3b4bddb7ab79f388a15009a24e63ebd50"}]}
PIPELINE-->

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151624/)
* Commit: 554bcdc2693b9d733728de501bef5df976935cf9
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-111243/554bcdc2693b9d733728de501bef5df976935cf9)
* This comment will update when the build is complete

### Failed CI Steps
* [Execute xpack-kibana-ciGroup6](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151624/execution/node/718/log)
* [Execute xpack-kibana-ciGroup6](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151624/execution/node/853/log)

### History
* :broken_heart: [Build #151310](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151310/) failed 7f9becd3b4bddb7ab79f388a15009a24e63ebd50
* :broken_heart: [Build #151151](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151151/) failed 48720733eb7d87ae23ccba82c2dd0ee161a87a59

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151151/","number":"151151","commit":"48720733eb7d87ae23ccba82c2dd0ee161a87a59"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151310/","number":"151310","commit":"7f9becd3b4bddb7ab79f388a15009a24e63ebd50"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151624/","number":"151624","commit":"554bcdc2693b9d733728de501bef5df976935cf9"}]}
PIPELINE-->

cjcenizal commented:
@alisonelizabeth and @sabarasaba, this is ready for your review!

cjcenizal commented on x-pack/plugins/upgrade_assistant/__jest__/client_integration/overview/fix_logs_step/fix_logs_step.test.tsx:
```diff
@@ -33,14 +33,49 @@ describe('Overview - Fix deprecation logs step', () => {
     server.restore();
   });
 
+  describe('Step status', () => {
```
@sabarasaba I added some logic that will set the step state to incomplete if the API poll receives `data` followed by an `error`. There's similar logic for the "Back up data" step, and I added a [test](https://github.com/elastic/kibana/pull/111243/files#diff-84d8cce1aad966faf177198b8104065830ec25a4d10f51fc9fbb7951b28f2e8fR160) to cover it. Could you add a similar test here in a followup PR?

sabarasaba replied:
Will do after we merge this one in 🔧

sabarasaba replied:
Added with https://github.com/elastic/kibana/pull/111701

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151671/)
* Commit: 768cd1ae39d6613ce414edd0d81cafd2327497a6
* Storybooks not built yet
* This comment will update when the build is complete

### Failed CI Steps
* [Check Types and Public API Docs](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151671/execution/node/428/log)

### History
* :broken_heart: [Build #151624](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151624/) failed 554bcdc2693b9d733728de501bef5df976935cf9
* :broken_heart: [Build #151310](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151310/) failed 7f9becd3b4bddb7ab79f388a15009a24e63ebd50
* :broken_heart: [Build #151151](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151151/) failed 48720733eb7d87ae23ccba82c2dd0ee161a87a59

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151151/","number":"151151","commit":"48720733eb7d87ae23ccba82c2dd0ee161a87a59"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151310/","number":"151310","commit":"7f9becd3b4bddb7ab79f388a15009a24e63ebd50"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151624/","number":"151624","commit":"554bcdc2693b9d733728de501bef5df976935cf9"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151671/","number":"151671","commit":"768cd1ae39d6613ce414edd0d81cafd2327497a6"}]}
PIPELINE-->

cjcenizal commented on x-pack/plugins/upgrade_assistant/__jest__/client_integration/helpers/time_manipulation.ts:
```diff
@@ -0,0 +1,24 @@
+export const advanceTime = async (ms: number) => {
+  await act(async () => {
+    jest.advanceTimersByTime(ms);
+    await flushPromiseJobQueue();
```
I think the flakiness in https://github.com/elastic/kibana/issues/111255 was due to not flushing this promise queue.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151681/)
* Commit: 0a14a6c2ac8a75735a0c152a9f5114a2810957ca
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-111243/0a14a6c2ac8a75735a0c152a9f5114a2810957ca)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/d4d52b06c2dd371950e46154a53ad9eda780f277) | [after](https://github.com/elastic/kibana/commit/0a14a6c2ac8a75735a0c152a9f5114a2810957ca) | diff |
| --- | --- | --- | --- |
| `upgradeAssistant` | 208.8KB | 210.9KB | +2.1KB |

### History
* :broken_heart: [Build #151671](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151671/) failed 768cd1ae39d6613ce414edd0d81cafd2327497a6
* :broken_heart: [Build #151624](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151624/) failed 554bcdc2693b9d733728de501bef5df976935cf9
* :broken_heart: [Build #151310](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151310/) failed 7f9becd3b4bddb7ab79f388a15009a24e63ebd50
* :broken_heart: [Build #151151](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151151/) failed 48720733eb7d87ae23ccba82c2dd0ee161a87a59

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151151/","number":"151151","commit":"48720733eb7d87ae23ccba82c2dd0ee161a87a59"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151310/","number":"151310","commit":"7f9becd3b4bddb7ab79f388a15009a24e63ebd50"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151624/","number":"151624","commit":"554bcdc2693b9d733728de501bef5df976935cf9"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151671/","number":"151671","commit":"768cd1ae39d6613ce414edd0d81cafd2327497a6"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151681/","number":"151681","commit":"0a14a6c2ac8a75735a0c152a9f5114a2810957ca"}]}
PIPELINE-->

cjcenizal commented:
@elasticmachine merge upstream

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151694/)
* Commit: 3340c42b347a165dbc695b9415c3628388e56711
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-111243/3340c42b347a165dbc695b9415c3628388e56711)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/cac2c82bea37abb68b134d8ca8b26811b6a82943) | [after](https://github.com/elastic/kibana/commit/3340c42b347a165dbc695b9415c3628388e56711) | diff |
| --- | --- | --- | --- |
| `upgradeAssistant` | 208.8KB | 210.9KB | +2.1KB |

### History
* :green_heart: [Build #151681](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151681/) succeeded 0a14a6c2ac8a75735a0c152a9f5114a2810957ca
* :broken_heart: [Build #151671](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151671/) failed 768cd1ae39d6613ce414edd0d81cafd2327497a6
* :broken_heart: [Build #151624](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151624/) failed 554bcdc2693b9d733728de501bef5df976935cf9
* :broken_heart: [Build #151310](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151310/) failed 7f9becd3b4bddb7ab79f388a15009a24e63ebd50
* :broken_heart: [Build #151151](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151151/) failed 48720733eb7d87ae23ccba82c2dd0ee161a87a59

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151151/","number":"151151","commit":"48720733eb7d87ae23ccba82c2dd0ee161a87a59"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151310/","number":"151310","commit":"7f9becd3b4bddb7ab79f388a15009a24e63ebd50"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151624/","number":"151624","commit":"554bcdc2693b9d733728de501bef5df976935cf9"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151671/","number":"151671","commit":"768cd1ae39d6613ce414edd0d81cafd2327497a6"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151681/","number":"151681","commit":"0a14a6c2ac8a75735a0c152a9f5114a2810957ca"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151694/","number":"151694","commit":"3340c42b347a165dbc695b9415c3628388e56711"}]}
PIPELINE-->

sabarasaba commented on x-pack/plugins/upgrade_assistant/public/application/components/overview/fix_logs_step/deprecations_count_checkpoint/deprecations_count_checkpoint.tsx:
```diff
@@ -73,10 +79,11 @@ export const DeprecationsCountCheckpoint: FunctionComponent = () => {
-  const calloutIcon = warningsCount > 0 ? 'alert' : 'check';
-  const calloutTestId = warningsCount > 0 ? 'hasWarningsCallout' : 'noWarningsCallout';
+  const logsCount = data?.count || 0;
+  const hasLogs = logsCount > 0;
```
👍🏼

sabarasaba reviewed (approved):
Latest changes lgtm 🚀

cjcenizal commented:
@elasticmachine merge upstream

alisonelizabeth commented on x-pack/plugins/upgrade_assistant/__jest__/client_integration/overview/fix_issues_step/fix_issues_step.test.tsx:
```diff
@@ -44,6 +44,37 @@ describe('Overview - Fix deprecation issues step', () => {
+      expect(exists(`fixIssuesStep-complete`)).toBe(true);
+    });
+
+    test('Its incomplete when there are critical deprecations', async () => {
```
```suggestion
    test(`It's incomplete when there are critical deprecations`, async () => {
```

alisonelizabeth commented on x-pack/plugins/upgrade_assistant/__jest__/client_integration/overview/fix_logs_step/fix_logs_step.test.tsx:
```diff
@@ -33,14 +33,49 @@ describe('Overview - Fix deprecation logs step', () => {
+      expect(exists(`fixLogsStep-complete`)).toBe(true);
+    });
+
+    test('Its incomplete when there are deprecation logs since last checkpoint', async () => {
```
```suggestion
    test(`It's incomplete when there are deprecation logs since last checkpoint`, async () => {
```

alisonelizabeth commented on x-pack/plugins/upgrade_assistant/public/application/components/overview/backup_step/cloud_backup.tsx:
```diff
@@ -20,22 +20,39 @@ import {
+      setIsComplete((!error && data?.isBackedUp) ?? false);
+    }
+
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Can you explain why this is needed?

alisonelizabeth commented on x-pack/plugins/upgrade_assistant/public/application/components/overview/fix_issues_step/fix_issues_step.tsx:
```diff
@@ -22,10 +23,39 @@ const i18nTexts = {
+}
+
+const FixIssuesStep: FunctionComponent<OverviewStepProps> = ({ setIsComplete }) => {
+  const [isEsFixed, setIsEsFixed] = useState(false);
```
[nit] It might be helpful to rename this or add a comment, indicating it representative of when there are 0 critical deprecations. It's clear to me now, but when I first saw `setIsFixed` as a prop in `es_stats.tsx` I had to pause a minute to think about what it represented, but maybe it's still too early for me 😅 .

cjcenizal replied:
👍  Sure, I added a comment. I think it makes sense to keep it abstract like this, so that we don't need to change function and prop names if we ever change the "is fixed" logic. For example, if we decide users need to fix all "warning" issues too.

alisonelizabeth commented on x-pack/plugins/upgrade_assistant/public/application/components/overview/fix_issues_step/fix_issues_step.tsx:
```diff
@@ -40,15 +70,7 @@ export const getFixIssuesStep = ({ nextMajor }: { nextMajor: number }): EuiStepP
-            <KibanaDeprecationStats />
-          </EuiFlexItem>
-        </EuiFlexGroup>
+        <FixIssuesStep isComplete={isComplete} setIsComplete={setIsComplete} />
```
it doesn't look like `FixIssuesStep` is using the `isComplete` prop

alisonelizabeth commented on x-pack/plugins/upgrade_assistant/public/application/components/overview/fix_logs_step/fix_logs_step.tsx:
```diff
@@ -88,17 +89,20 @@ const FixLogsStep: FunctionComponent = () => {
           </EuiText>
           <EuiSpacer size="m" />
-          <DeprecationsCountCheckpoint />
+          <DeprecationsCountCheckpoint setHasNoDeprecationLogs={setIsComplete!} />
```
Is the `!` necessary here?

alisonelizabeth reviewed (commented):
Thanks for working on this @cjcenizal! Overall LGTM. I am curious about the usage of `// eslint-disable-next-line react-hooks/exhaustive-deps` throughout the code and if it can be avoided. Let me know what you think.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151804/)
* Commit: 0e2dc7f9874b0d8fd179b38aa54d63d7f2386d68
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-111243/0e2dc7f9874b0d8fd179b38aa54d63d7f2386d68)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/8978d1c3345a8f4b2849ea1e10626ec0d81533f4) | [after](https://github.com/elastic/kibana/commit/0e2dc7f9874b0d8fd179b38aa54d63d7f2386d68) | diff |
| --- | --- | --- | --- |
| `upgradeAssistant` | 208.8KB | 210.9KB | +2.1KB |

### History
* :green_heart: [Build #151694](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151694/) succeeded 3340c42b347a165dbc695b9415c3628388e56711
* :green_heart: [Build #151681](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151681/) succeeded 0a14a6c2ac8a75735a0c152a9f5114a2810957ca
* :broken_heart: [Build #151671](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151671/) failed 768cd1ae39d6613ce414edd0d81cafd2327497a6
* :broken_heart: [Build #151624](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151624/) failed 554bcdc2693b9d733728de501bef5df976935cf9
* :broken_heart: [Build #151310](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151310/) failed 7f9becd3b4bddb7ab79f388a15009a24e63ebd50

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151310/","number":"151310","commit":"7f9becd3b4bddb7ab79f388a15009a24e63ebd50"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151624/","number":"151624","commit":"554bcdc2693b9d733728de501bef5df976935cf9"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151671/","number":"151671","commit":"768cd1ae39d6613ce414edd0d81cafd2327497a6"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151681/","number":"151681","commit":"0a14a6c2ac8a75735a0c152a9f5114a2810957ca"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151694/","number":"151694","commit":"3340c42b347a165dbc695b9415c3628388e56711"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151804/","number":"151804","commit":"0e2dc7f9874b0d8fd179b38aa54d63d7f2386d68"}]}
PIPELINE-->

cjcenizal commented:
@elasticmachine merge upstream

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151886/)
* Commit: 7d3e404dcad76a1c0be13a5e567f938c616936ce
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-111243/7d3e404dcad76a1c0be13a5e567f938c616936ce)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/c2a189e9d15653065ed22daa394876040dc9fc61) | [after](https://github.com/elastic/kibana/commit/7d3e404dcad76a1c0be13a5e567f938c616936ce) | diff |
| --- | --- | --- | --- |
| `upgradeAssistant` | 208.8KB | 210.9KB | +2.1KB |

### History
* :green_heart: [Build #151804](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151804/) succeeded 0e2dc7f9874b0d8fd179b38aa54d63d7f2386d68
* :green_heart: [Build #151694](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151694/) succeeded 3340c42b347a165dbc695b9415c3628388e56711
* :green_heart: [Build #151681](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151681/) succeeded 0a14a6c2ac8a75735a0c152a9f5114a2810957ca
* :broken_heart: [Build #151671](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151671/) failed 768cd1ae39d6613ce414edd0d81cafd2327497a6
* :broken_heart: [Build #151624](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151624/) failed 554bcdc2693b9d733728de501bef5df976935cf9

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151624/","number":"151624","commit":"554bcdc2693b9d733728de501bef5df976935cf9"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151671/","number":"151671","commit":"768cd1ae39d6613ce414edd0d81cafd2327497a6"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151681/","number":"151681","commit":"0a14a6c2ac8a75735a0c152a9f5114a2810957ca"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151694/","number":"151694","commit":"3340c42b347a165dbc695b9415c3628388e56711"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151804/","number":"151804","commit":"0e2dc7f9874b0d8fd179b38aa54d63d7f2386d68"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151886/","number":"151886","commit":"7d3e404dcad76a1c0be13a5e567f938c616936ce"}]}
PIPELINE-->

cjcenizal commented:
@alisonelizabeth Thanks for your review! Could you please take another look? I addressed your feedback, including implementing memoization in https://github.com/elastic/kibana/pull/111243/commits/7ccb6abfc75dacf5a6fbfb8b743715265eaa2f3f.

alisonelizabeth commented on x-pack/plugins/upgrade_assistant/public/application/components/overview/fix_issues_step/es_stats/es_stats.tsx:
```diff
@@ -35,18 +35,34 @@ const i18nTexts = {
+    if (!isLoading && !error) {
+      setIsFixed(criticalDeprecations.length === 0);
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Is this still needed?

alisonelizabeth commented on x-pack/plugins/upgrade_assistant/public/application/components/overview/fix_issues_step/kibana_stats/kibana_stats.tsx:
```diff
@@ -77,6 +81,14 @@ export const KibanaDeprecationStats: FunctionComponent = () => {
+      setIsFixed(criticalDeprecationsCount === 0);
+    }
+
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Is this still needed?

alisonelizabeth reviewed (approved):
Thanks for addressing my feedback! I noticed two more places where `// eslint-disable-next-line react-hooks/exhaustive-deps` is being applied. Would you mind taking a quick look before merging? Otherwise LGTM.

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151955/)
* Commit: b5b082bb9d3e31927452466403adc6c86ffb903f
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-111243/b5b082bb9d3e31927452466403adc6c86ffb903f)
* This comment will update when the build is complete

### Failed CI Steps
* [Jest Unit Tests](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151955/execution/node/190/log)

### History
* :green_heart: [Build #151886](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151886/) succeeded 7d3e404dcad76a1c0be13a5e567f938c616936ce
* :green_heart: [Build #151804](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151804/) succeeded 0e2dc7f9874b0d8fd179b38aa54d63d7f2386d68
* :green_heart: [Build #151694](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151694/) succeeded 3340c42b347a165dbc695b9415c3628388e56711
* :green_heart: [Build #151681](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151681/) succeeded 0a14a6c2ac8a75735a0c152a9f5114a2810957ca
* :broken_heart: [Build #151671](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151671/) failed 768cd1ae39d6613ce414edd0d81cafd2327497a6

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151671/","number":"151671","commit":"768cd1ae39d6613ce414edd0d81cafd2327497a6"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151681/","number":"151681","commit":"0a14a6c2ac8a75735a0c152a9f5114a2810957ca"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151694/","number":"151694","commit":"3340c42b347a165dbc695b9415c3628388e56711"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151804/","number":"151804","commit":"0e2dc7f9874b0d8fd179b38aa54d63d7f2386d68"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151886/","number":"151886","commit":"7d3e404dcad76a1c0be13a5e567f938c616936ce"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151955/","number":"151955","commit":"b5b082bb9d3e31927452466403adc6c86ffb903f"}]}
PIPELINE-->

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151972/)
* Commit: f40b2507c63aab495b928709a4f49d386ac417b4
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-111243/f40b2507c63aab495b928709a4f49d386ac417b4)
* This comment will update when the build is complete

### Failed CI Steps
* [Jest Unit Tests](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151972/execution/node/190/log)

### History
* :broken_heart: [Build #151955](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151955/) failed b5b082bb9d3e31927452466403adc6c86ffb903f
* :green_heart: [Build #151886](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151886/) succeeded 7d3e404dcad76a1c0be13a5e567f938c616936ce
* :green_heart: [Build #151804](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151804/) succeeded 0e2dc7f9874b0d8fd179b38aa54d63d7f2386d68
* :green_heart: [Build #151694](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151694/) succeeded 3340c42b347a165dbc695b9415c3628388e56711
* :green_heart: [Build #151681](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151681/) succeeded 0a14a6c2ac8a75735a0c152a9f5114a2810957ca

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151681/","number":"151681","commit":"0a14a6c2ac8a75735a0c152a9f5114a2810957ca"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151694/","number":"151694","commit":"3340c42b347a165dbc695b9415c3628388e56711"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151804/","number":"151804","commit":"0e2dc7f9874b0d8fd179b38aa54d63d7f2386d68"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151886/","number":"151886","commit":"7d3e404dcad76a1c0be13a5e567f938c616936ce"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151955/","number":"151955","commit":"b5b082bb9d3e31927452466403adc6c86ffb903f"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151972/","number":"151972","commit":"f40b2507c63aab495b928709a4f49d386ac417b4"}]}
PIPELINE-->

cjcenizal commented:
@alisonelizabeth It turns out that defining exhaustive deps creates an infinite loop:
1. In the overview, `completedStepsMap` is created and a new `setCompletedStep` is derived from it.
2. `setCompletedStep` is passed to a step.
3. In the step, `useEffect` is triggered by the provided `setCompletedStep`.
4. If the step calls `setCompletedStep`, then a new `completedStepsMap` is created in the overview. At this point we're back at step 1 again, and we're stuck in an infinite loop.

I'm not sure what the best way out of this is. I suppose the `useEffect` could evaluate the current completion state, and only call `setCompletedStep` if it determines that the state should change from `true` to `false` or vice versa. For now I think it's best to selectively disable the exhaustive deps rule. I've added comments to explain why. Do you mind if we merge as-is and figure out the best way to refactor this separately?

alisonelizabeth commented:
>  For now I think it's best to selectively disable the exhaustive deps rule. I've added comments to explain why. Do you mind if we merge as-is and figure out the best way to refactor this separately?

👍 That's fine with me. I do think it's worth revisiting, but happy to have this merged as-is with comments.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152023/)
* Commit: 416aa3bb0d2c4301d1334ff3e3b40a01db0da577
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-111243/416aa3bb0d2c4301d1334ff3e3b40a01db0da577)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/e1ae48578f1ba1a73476a3d9c2be8d845b3055a4) | [after](https://github.com/elastic/kibana/commit/416aa3bb0d2c4301d1334ff3e3b40a01db0da577) | diff |
| --- | --- | --- | --- |
| `upgradeAssistant` | 205.9KB | 208.1KB | +2.1KB |

### History
* :broken_heart: [Build #151972](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151972/) failed f40b2507c63aab495b928709a4f49d386ac417b4
* :broken_heart: [Build #151955](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151955/) failed b5b082bb9d3e31927452466403adc6c86ffb903f
* :green_heart: [Build #151886](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151886/) succeeded 7d3e404dcad76a1c0be13a5e567f938c616936ce
* :green_heart: [Build #151804](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151804/) succeeded 0e2dc7f9874b0d8fd179b38aa54d63d7f2386d68
* :green_heart: [Build #151694](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151694/) succeeded 3340c42b347a165dbc695b9415c3628388e56711

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151694/","number":"151694","commit":"3340c42b347a165dbc695b9415c3628388e56711"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151804/","number":"151804","commit":"0e2dc7f9874b0d8fd179b38aa54d63d7f2386d68"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151886/","number":"151886","commit":"7d3e404dcad76a1c0be13a5e567f938c616936ce"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151955/","number":"151955","commit":"b5b082bb9d3e31927452466403adc6c86ffb903f"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151972/","number":"151972","commit":"f40b2507c63aab495b928709a4f49d386ac417b4"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152023/","number":"152023","commit":"416aa3bb0d2c4301d1334ff3e3b40a01db0da577"}]}
PIPELINE-->
